src/stream/src/common/table/state_table.rs

```diff
@@ -1096,43 +1092,53 @@ where
         vnode: VirtualNode,
         prefetch_options: PrefetchOptions,
     ) -> StreamExecutorResult<RowStream<'_, S, SD>> {
-        Ok(self
-            .iter_key_and_val_with_pk_range(pk_range, vnode, prefetch_options)
-            .await?
-            .map(get_second))
+        Ok(deserialize_row_stream(
+            self.iter_kv_with_pk_range(pk_range, vnode, prefetch_options)
+                .await?,
+            &self.row_serde,
+        ))
     }
 
-    pub async fn iter_key_and_val_with_pk_range(
+    /// This function scans rows from the relational table with specific `pk_range` under the same
+    /// `vnode`. It returns both the key with vnode and the row.
+    pub async fn iter_row_and_key_with_pk_range(
         &self,
         pk_range: &(Bound<impl Row>, Bound<impl Row>),
         // Optional vnode that returns an iterator only over the given range under that vnode.
         // For now, we require this parameter, and will panic. In the future, when `None`, we can
         // iterate over each vnode that the `StateTableInner` owns.
         vnode: VirtualNode,
         prefetch_options: PrefetchOptions,
-    ) -> StreamExecutorResult<RowStreamWithPk<'_, S, SD>> {
-        Ok(deserialize_row_stream(
-            self.iter_with_pk_range_inner(pk_range, vnode, prefetch_options)
+    ) -> StreamExecutorResult<RowStreamWithKey<'_, S, SD>> {
+        Ok(deserialize_row_stream_with_key(
+            self.iter_kv_with_pk_range(pk_range, vnode, prefetch_options)
                 .await?,
             &self.row_serde,
         ))
     }
 
-    /// This function scans rows from the relational table with specific `pk_prefix`, return both
-    /// key and value.
-    pub async fn iter_key_and_val(
+    async fn iter_kv(
         &self,
-        pk_prefix: impl Row,
+        key_range: (Bound<Bytes>, Bound<Bytes>),
+        prefix_hint: Option<Bytes>,
         prefetch_options: PrefetchOptions,
-    ) -> StreamExecutorResult<RowStreamWithPk<'_, S, SD>> {
-        Ok(deserialize_row_stream(
-            self.iter_with_pk_prefix_inner(pk_prefix, prefetch_options)
-                .await?,
-            &self.row_serde,
-        ))
+    ) -> StreamExecutorResult<<S::Local as LocalStateStore>::IterStream<'_>> {
+        let read_options = ReadOptions {
+            prefix_hint,
+            ignore_range_tombstone: false,
+            retention_seconds: self.table_option.retention_seconds,
+            table_id: self.table_id,
+            read_version_from_backup: false,
+            prefetch_options,
+            cache_policy: CachePolicy::Fill(CachePriority::High),
+        };
+
+        Ok(self.local_store.iter(key_range, read_options).await?)
     }
 
-    async fn iter_with_pk_prefix_inner(
+    /// This function scans raw key-values from the relational table with specific `pk_prefix`.
+    /// `pk_prefix` is used to identify the exact vnode the scan should perform on.
```

Comment:
Should we add a comment here to warn users that `raw key-values` means encoded bytes of `vnode` will be a prefix of the key?

Perhaps we should also add a new interface (can be a separate PR) that iterates over `pk` and `row`, without `vnode`, so we can use it in `merge_sort`.

Otherwise correctness of `merge_sort` is affected, because as mentioned here https://github.com/risingwavelabs/risingwave/issues/11869, we will be sorting on the vnode prefix too.

Reply:
I guess we can prevent developers from misusing these interfaces once and for all with the help of the type system. Like `FullKey<Bytes>`, we can also have `TableKey` or `RealKey` as the return types. We may even unify all these interfaces with the generic of `impl FromRawKv`.

```rust
pub trait FromRawKv {
  fn from_raw(key: FullKey<Bytes>, value: Bytes) -> Self;
}

impl FromRawKv for OwnedRow { .. }
impl FromRawKv for (TableKey, OwnedRow) { .. }
impl FromRawKv for (RealKey, OwnedRow) { .. }

impl StateTable {
  fn iter<R: FromRawKv>(..) -> impl TryStream<R> { .. }
}
```

Reply:
> Otherwise correctness of merge_sort is affected, because as mentioned here https://github.com/risingwavelabs/risingwave/issues/11869, we will be sorting on the vnode prefix too.

I thought the merge_sort is only used to merge data from different vnodes. Anyway, now that the interface returns a tuple (VirtualNode, Bytes) instead of raw Bytes as the key so I think it is clear enough. If one wants to merge across vnodes, the returning stream can be used directly. If one wants to merge by ignoring vnodes, the returning stream key can be mapped to ignore the VirtualNode part of the key before doing merge_sort.

Reply:
> I guess we can prevent developers from misusing these interfaces once and for all with the help of the type system. Like `FullKey<Bytes>`, we can also have `TableKey` or `RealKey` as the return types. We may even unify all these interfaces with the generic of `impl FromRawKv`.
> 
> ```rust
> pub trait FromRawKv {
>   fn from_raw(key: FullKey<Bytes>, value: Bytes) -> Self;
> }
> 
> impl FromRawKv for OwnedRow { .. }
> impl FromRawKv for (TableKey, OwnedRow) { .. }
> impl FromRawKv for (RealKey, OwnedRow) { .. }
> 
> impl StateTable {
>   fn iter<R: FromRawKv>(..) -> impl TryStream<R> { .. }
> }
> ```

Good idea. Let me try.